pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,197 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitortool
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"github.com/go-ping/ping"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+)
+
+// MonitorTool is a tool to monitor the latency of the node.
+type MonitorTool struct {
+	latencyStore *LatencyStore
+
+	interval time.Duration
+	timeout  time.Duration
+}
+
+// TODO: Maybe we need to implement it like podStore.
+// Simple nodeStore struct
+type NodeStore struct {
+	mutex     sync.RWMutex
+	PingItems map[string]PingItem
+}
+
+// TODO: NodeInternalIP/NodeExternalIP
+// We only need to store the NodeInternalIP of the node.
+// In first step, we only use the nodeip tracker to get the node internal/external IP.
+type PingItem struct {
+	// Name is the name of the node.
+	Name string
+	// IP is the IP of the node.
+	IPs []string
+}
+
+func NewNodeStore() *NodeStore {
+	return &NodeStore{
+		PingItems: make(map[string]PingItem),
+	}
+}
+
+func (n *NodeStore) Clear() {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems = make(map[string]PingItem)
+}
+
+func (n *NodeStore) AddPingItem(name string, ips []string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems[name] = PingItem{
+		Name: name,
+		IPs:  ips,
+	}
+}
+
+func (n *NodeStore) GetPingItem(name string) (PingItem, bool) {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	item, found := n.PingItems[name]
+	return item, found
+}
+
+func (n *NodeStore) DeletePingItem(name string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	delete(n.PingItems, name)
+}
+
+func (n *NodeStore) ListPingItems() []PingItem {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	items := make([]PingItem, 0, len(n.PingItems))
+	for _, item := range n.PingItems {
+		items = append(items, item)
+	}
+	return items
+}
+
+func (n *NodeStore) UpdatePingItem(name string, ips []string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems[name] = PingItem{
+		Name: name,
+		IPs:  ips,
+	}
+}
+
+func NewNodeLatencyMonitor(nodeInformer coreinformers.NodeInformer, interval, timeout time.Duration) *MonitorTool {
+	return &MonitorTool{
+		latencyStore: NewLatencyStore(nodeInformer),
+		interval:     interval,
+		timeout:      timeout,
+	}
+}
+
+func (m *MonitorTool) pingAll() {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+
+	// TODO: Get current node internal/external IP.
+	fromIP := ""
+
+	klog.InfoS("Start to ping all nodes")
+	wg := sync.WaitGroup{}
+	for toIP, name := range nodeIPs {
+		wg.Add(1)
+
+		// TODO: Add ping limiter
+		go func(toIP string, name string) {
```

Comment:
I feel like it would be better if we had a bounded number of goroutines doing this, instead of creating one per node.

Reply:
I'll use ratelimter to limit the number of goroutines, and it can be configured in config.